src/mbgl/renderer/image_manager.cpp

```diff
@@ -73,14 +80,57 @@ void ImageManager::getImages(ImageRequestor& requestor, ImageRequestPair&& pair)
         }
     }
     if (isLoaded() || hasAllDependencies) {
-        notify(requestor, std::move(pair));
+        checkMissingAndNotify(requestor, std::move(pair));
     } else {
         requestors.emplace(&requestor, std::move(pair));
     }
 }
 
 void ImageManager::removeRequestor(ImageRequestor& requestor) {
     requestors.erase(&requestor);
+    missingImageRequestors.erase(&requestor);
+}
+
+void ImageManager::imagesAdded() {
+    for (auto it = missingImageRequestors.begin(); it != missingImageRequestors.end();) {
+        if (it->second.callbacksRemaining == 0) {
+            notify(*it->first, it->second.pair);
+            missingImageRequestors.erase(it++);
+        } else {
+            it++;
+        }
+    }
+}
+
+void ImageManager::checkMissingAndNotify(ImageRequestor& requestor, const ImageRequestPair& pair) {
+    int missing = 0;
+    for (const auto& dependency : pair.first) {
+        auto it = images.find(dependency.first);
+        if (it == images.end()) {
+            missing++;
+        }
+    }
+
+
+    if (missing > 0) {
+        ImageRequestor* requestorPtr = &requestor;
+        missingImageRequestors.emplace(requestorPtr, MissingImageRequestPair { std::move(pair), missing });
+
+        for (const auto& dependency : pair.first) {
+            auto it = images.find(dependency.first);
+            if (it == images.end()) {
+                observer->onStyleImageMissing(dependency.first, [this, requestorPtr]() {
+                    auto requestorIt = missingImageRequestors.find(requestorPtr);
```

Comment:
Should the callback have a return value to indicate whether or not an image was added to satisfy the missing id? That way `missingImageRequestors` is not always tracking missing images, especially useful in cases where there is no intention to add the missing images in response to `onStyleImageMissing`.

Reply:
On the next frame `ImageManager#imagesAdded()` cleans this up so that it isn't tracking them forever. What would the return value help with it? I could see it allowing the tile parsing to start a fraction of a frame earlier but is that worth it?

Reply:
Yeah, return value doesn't make sense. I see that the diff checking in `renderer::impl` takes care of any benefit that would be had from triggering an update.